rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+
+1. There are a number of types of sinks that can differ in implementation details
+2. We provide a simple, unified, stream-based interface to drive all of them
+
+The most important benefit here is that we don't need to rewrite everything at
+once to move to a new style. You can simply wrap `Sink::OldAndBad` around
+existing implementation and upgrade them later one at a time.
+
+Another important point is that we've decoupled the topology-level API from the
+implementation-level API. This means we can start with something simple like the
+`run` method above, and later move towards something more sophisticated without
+necessarily touching every individual component.
+
+Once this seam is established, the next most important abstraction is
+`ServiceSink`. We have something similar right now in `BatchSink`, but we've
+begun to see limitations of that design. A full proposal is outside of the scope
+of this RFC, but future work should explore the relationship between batching,
+encoding, and request building. Their separation in the current design appears
+to be the cause of some limitations and there's likely an improvement to be made
+by evaluating and adjusting to their mutual dependencies.
+
+### Topology design
+
+With the above component changes in place, we'll gain some flexibility in how we
+build topologies. The goal will be, where possible, to consolidate stateless
+transform-like processing into cloneable `Push` implementations that are passed
+to sources in place of the current raw channels senders. This will effectively
+"inline" this logic into the source task, allowing it to execute with the
+natural concurrency of a given source (e.g. per connection).
+
+#### Runtime task layout
+
+For example, consider a Vector configuration with a TCP syslog source, grok
+parser, rename fields transform, and Cloudwatch logs sink. With our existing
+design, each of those components exists as a single independent task, connected
+together with channels. While simple, this leads to the whole pipeline being
+limited by the single-threaded throughput of the slowest component.
+
+In the new design, we take advantage of the fact that all of the transform work
+for this config is embarrassingly parallel. Where the source would normally
+clone a channel `Sender` for each incoming connection, it will now clone
+a `Push` implementation called `Pipeline` (naming TBD) that does the
+transformation work inline before forwarding the results directly to the sink.
+This allows our CPU concurrency to scale up naturally with the natural
+concurrency of the source.
+
+In an abstract sense, the goal of this change is to transition from a topology
+of individual, heterogeneous tasks to one of many consolidated, homogeneous
+tasks. Take for example the below topology:
+
+![Example topology](./2020-06-18-2625-architecture-revisit/combined.png)
+
+By shifting the stateless transforms up into the connection handling tasks, we
```

Comment:
I feel like this section would be strengthened if it talked about how ordering would be managed and kept efficient.

Reply:
I'm not sure I follow. The idea of moving stateless transforms is that they are explicitly unaffected by ordering, and therefore we do not need to manage it. Unless you mean stating that more clearly :smile:

Reply:
So does this mean they might provide output to the sink in an unordered way even though their input is ordered?

Reply:
There is no difference in the logical ordering of events as received by sinks.

Currently, events received/created concurrently on different source tasks are serialized arbitrarily into a single output channel before flowing through subsequent transforms. With this change, they'll flow through transforms on the source task before being arbitrarily serialized in exactly the same way. The transforms being inlined are not capable of reordering events within their task. The only difference is that events from different tasks remain logically concurrent for longer, which could allow for slight timing differences across tasks. But there are already no ordering guarantees across different source tasks.